### PR TITLE
batches: track fork status on changeset specs

### LINKED ISSUE #29632: Forks support testing
- Changesets will not indicate a fork when it's still processing - this could be confusing.

image

- Indicator on preview page is lost when changeset is still on a fork but flag is disabled (maybe should also look at `ExternalForkNamespace`)

image

- Error after turning off the flag and updating a forked changeset:

> [enterprise-repo-updater] WARN Marked record as errored, name: batches_reconciler_worker, id: 20339, err: forking repository: request to http://127.0.0.1:3180/repos/sourcegraph/automation-testing/forks returned status 422: 'eseliger' is the login for a user account. You must pass the login for an organization account.

### PR DESCRIPTION
**Pull request by LawnGnome**
This is required to be able to display the fork indicator consistently when previewing and processing new changesets, and moves the place we check the global fork state out of the reconciler, which is probably more correct.

As a bonus, this will greatly simplify adding new fields to the changeset spec that control fork behaviour, since we can simply extend `computeForkNamespace()` to take that into account.

One wrinkle here is that changesets that haven't yet been published to the user fork have to have some way of indicating that they're going to be pushed to a user fork. We don't know the credential that will eventually be used, so we don't know the namespace name. As a result, I've introduced a new extension to the branch UI changes that were made in #29367:

image

Fixes #29632.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/batches.graphql [hunk 1]
```diff
@@ -840,6 +840,13 @@ type VisibleChangesetSpec implements ChangesetSpec & Node {
     """
     expiresAt: DateTime
 
+    """
+    The fork namespace the changeset spec should be pushed to, if any. The
+    special value "<user>" is used to indicate that the changeset spec should be
+    pushed to the namespace associated with the user or site credential in use.
+    """
+    forkNamespace: String
```
[the remark this answers is not in the document]
Comment (LawnGnome) [reply cmd/frontend/graphqlbackend/batches.graphql:848]: I'm definitely open to suggestions on this. I'd probably represent this as an `enum` in Rust or a union type in TypeScript, but I don't know the best way to translate that to GraphQL.

I don't particularly want to have a boolean `fork` on specs, though: it's likely in the future that we'll support specifying arbitrary fork namespaces, and that would become a dead end that would have to be deprecated.

One option I did consider was adding something like:

```graphql
interface ForkTarget {
  user: Boolean!
  namespace: String
}
```

And then making `forkNamespace` that type.

[the remark this answers is not in the document]
Comment (eseliger) [reply cmd/frontend/graphqlbackend/batches.graphql:848]: For the same of getting it into the release: How about we keep it as is and add the experimental note on just this very field so we can think about it a bit more for the next release? Doesn't need to be the entire feature, just this single field.

> Arbitrary fork namespaces

Good idea. What would user signify in your proposal?

[the remark this answers is not in the document]
Comment (LawnGnome) [reply cmd/frontend/graphqlbackend/batches.graphql:848]: I'm mostly done with implementing this, so let's see if we can sneak it in.

>Good idea. What would user signify in your proposal?

If we were pushing to a fork in namespace `org`, then that type would return something like:

```
user: false
namespace: "org"
```

[the remark this answers is not in the document]
Comment (eseliger) [reply cmd/frontend/graphqlbackend/batches.graphql:848]: Would we be able to tell it's not a user before we actually forked it? 🤔 From an arbitrary string it seems impossible to derive if it's a user or org on the codehost 🤔 
Regardless, seems like you got a solid plan here, don't want to derail this further before a release cut :)

[the remark this answers is not in the document]
Comment (LawnGnome) [reply cmd/frontend/graphqlbackend/batches.graphql:848]: Ah, I see the confusion, sorry. `user` doesn't mean "any user"; it's "the user associated with the push credential".

[the remark this answers is not in the document]
Comment (LawnGnome) [reply cmd/frontend/graphqlbackend/batches.graphql]: For the record, I called it `pushUser` in the actual new schema type.

**Pull request merged**